packages/context/src/binding.ts

```diff
@@ -95,6 +99,41 @@ export class Binding {
   }
 
   /**
+   * Bind the key to a BindingProvider
+   */
+  public toProvider<T>(providerClass: Constructor<Provider<T>>): this {
+    this.getValue = async (ctx): Promise<BoundValue> => {
+      const providerInstance: Provider<T> = await this.getProviderInstance(ctx, providerClass);
```

Comment:
Ideally, we should detect whether `this.getProviderInstance` returned a promise or a provide value. When `getProviderInstance` returned the value synchronously, `this.getValue` should return synchronously too. This is how other `getValue` functions behave now.

Feel free to address this comment later, after this initial PR is landed.

Here is what I am thinking about - notice how much simpler and shorter the code becomes:

```ts
public toProvider<T>(providerClass: Constructor<Provider<T>>): this {
  this.getValue = ctx => {
    const providerOrPromise = instantiateClass<Provider<T>>(providerClass, ctx);
    if (isPromise(providerOrPromise)) {
      return providerOrPromise.then(p => p.value(ctx));
    } else {
      return providerOrPromise.value(ctx);
    }
  }
}
```